File: pressingKeypad.cpp
#include <bits/stdc++.h>
using namespace std;

string solution(vector<int> numbers, string hand) {
    string answer = "";
    if (hand == "right")
        hand = 'R';
    else
        hand = 'L';
    for (auto& i : numbers) {
        if (i == 0)
            i = 10;
        else
            i--;
    }
    int left = 9;
    int right = 11;
    for (auto& i : numbers) {
        if (i == 0 || i == 3 || i == 6) {
            answer += 'L';
            left = i;
            cout << "i: " << i << '\n';
        } else if (i == 2 || i == 5 || i == 8) {
            answer += 'R';
            right = i;
            cout << "i: " << i << '\n';
        } else {
            int l_d = abs(left / 3 - i / 3) + abs(left % 3 - i % 3);
            int r_d = abs(right / 3 - i / 3) + abs(right % 3 - i % 3);
            cout << "i: " << i << " l_d: " << l_d << " r_d: " << r_d << '\n';
            if (l_d == r_d) {
                answer += hand;
                string R = "R";
                if (hand == R)
                    right = i;
                else
                    left = i;
            } else if (l_d < r_d) {
                answer += 'L';
                left = i;
            } else {
                answer += 'R';
                right = i;
            }
        }
    }
    return answer;
}

File: pressingKeypad.js
function solution(numbers, hand) {
    let answer = "";
    let lx = 3,
        ly = 0,
        rx = 3,
        ry = 2;
    for (let i = 0; i < numbers.length; i++) {
        if (numbers[i] === 0) {
            numbers[i] = 11;
        }
        numbers[i]--;
        let x = Math.floor(numbers[i] / 3);
        let y = numbers[i] % 3;
        let ldist = Math.abs(x - lx) + Math.abs(y - ly);
        let rdist = Math.abs(x - rx) + Math.abs(y - ry);
        let nb = numbers[i];
        if (nb === 0 || nb === 3 || nb === 6) {
            answer += "L";
            lx = x;
            ly = y;
        } else if (nb === 2 || nb === 5 || nb === 8) {
            answer += "R";
            rx = x;
            ry = y;
        } else {
            if (ldist < rdist) {
                answer += "L";
                lx = x;
                ly = y;
            } else if (ldist > rdist) {
                answer += "R";
                rx = x;
                ry = y;
            } else {
                if (hand === "left") {
                    answer += "L";
                    lx = x;
                    ly = y;
                } else {
                    answer += "R";
                    rx = x;
                    ry = y;
                }
            }
        }
    }
    return answer;
}
